cmd/thanos/query.go

```diff
@@ -622,9 +623,25 @@ func runQuery(
 		)
 
 		g.Add(func() error {
+			// Wait for initial endpoint update before marking as ready
+			// Use store response timeout as timeout, if not set, use 30 seconds as default
+			timeout := storeResponseTimeout
+			if timeout == 0 {
+				timeout = 30 * time.Second
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+
+			level.Info(logger).Log("msg", "waiting for initial endpoint discovery before marking gRPC as ready", "timeout", timeout)
+			if err := endpointSet.WaitForFirstUpdate(ctx); err != nil {
```

Comment:
we still mark as ready if it fails, shouldnt we wrap into "runutil.Retry"?

Reply:
Doesnt it make more sense to just call "Update" here instead of waiting for another place to call it?

Reply:
> we still mark as ready if it fails, shouldnt we wrap into "runutil.Retry"?
That is a valid point, but right now we dont even return an error from Update(). So what should we consider an error here? 
Maybe this is something for the `healthy` prober?

Reply:
> Doesnt it make more sense to just call "Update" here instead of waiting for another place to call it?

By this point, we already called Update(), but we are just forcing waiting here. We could instead call it blocking here, but it would just cause contention (over the lock in EndpointSet) and a bunch of duplicated GRPC calls, in our case we have hundreds of Receivers, for example.

I am fine with both, but I would rather stick to this solution. As it makes more sense from a performance PoV, since calling it blocking here as well will force a second serial call to `Update()`, after the one in `setupEndpointSet` finishes (it is using `runutil.Repeat(`, which runs once and then repeats).

Reply:
@MichaHoffmann does this make sense? Do you still have some comments/suggestions/etc.?